# Group 22 Project Final Report

*An Analysis of the Risk of Developing Heart Disease for Individuals in Cleveland, Ohio.*

## Introduction

Cardiovascular disease includes any condition that affects the circulatory system. This results in negative impacts on the function of the heart and veins, which differ from different types of cardiovascular disease (Thiriet, 1970). Cardiovascular disease is also commonly known as heart disease (Thiriet, 1970). 
In order to save many lives, it is essential that individuals with heart disease are diagnosed early to start treatment as soon as possible (Pal et al., 2022). A model that can accurately predict the risk of developing cardiovascular disease would have the potential to save countless lives (Pal et al., 2022). 
By narrowing it down to specific factors such as cholesterol or blood pressure we can possibly give advice on proactive measures that are most important to minimize risk of heart disease.

Therefore the question we intend to answer is: Can we predict which patients in Cleveland, Ohio are at the highest risk of developing heart disease?

The dataset we will be using was retrieved from "processed.cleveland.data" in the Heart Disease Data Set (https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/) provided by UCI Machine Learning Repository. The dataset includes the following 14 attributes which can be numerical or categorical.

## Summary of Preliminary Analysis

We will start with a quick summary of our previous analysis. Our table describes which variables have apossible correlation with heart disease. We noticed some minor mistakes in our analysis but we will proceed with 2 numerical variables than seem to have correlation with heart disease

### Loading, Wrangling, and Cleaning the Data

In [1]:
# This cleaning-up procedure is mostly the same as our proposal
# We added some steps to omit rows containing unclassified values
import altair as alt
import numpy as np
import pandas as pd
from sklearn.compose import make_column_transformer
from sklearn.preprocessing import StandardScaler

In [2]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"

original_data = pd.read_csv(url, names=[
    "age", 
    "sex", 
    "chest_pain", 
    "resting_blood_pressure", 
    "cholesterol", 
    "fasting_blood_sugar", 
    "resting_electrocardiographic_results", 
    "max_heart_rate",
    "exercise_induced_angina",  
    "st_depression_exercise", 
    "slope_st", 
    "major_vessels", 
    "thal",
    "diagnosis"
])

hd_data = original_data.dropna(how='all')

# omit rows containing unclassified values
hd_data = hd_data.drop(
    hd_data[hd_data.major_vessels == '?'].index
)
hd_data["major_vessels"] = pd.to_numeric(hd_data["major_vessels"])

hd_data = hd_data.drop(
    hd_data[hd_data.thal == '?'].index
)

In [3]:
# replace numeric values with categorical lables for categorical variables 
# except for 'resting_electrocardiographic_results' which requires a long description for each label
hd_data["sex"] = hd_data["sex"].apply(lambda x: "male" if (x == 1.0) else "female")

hd_data["chest_pain"] = hd_data["chest_pain"].replace({
    1: 'Typical angina',
    2: 'Atypical angina',
    3: 'Non-anginal pain',
    4: 'Asymptomatic'
})

hd_data["fasting_blood_sugar"] = hd_data["fasting_blood_sugar"].apply(lambda x: True if (x == 1.0) else False)

hd_data["exercise_induced_angina"] = hd_data["exercise_induced_angina"].apply(lambda x: 'Yes' if (x == 1.0) else 'No')

hd_data["slope_st"] = hd_data["slope_st"].replace({
    1.0: 'Upsloping',
    2.0: 'Flat',
    3.0: 'Downsloping'
})

hd_data["thal"] = hd_data["thal"].replace({
    '3.0': 'Normal',
    '6.0': 'Fixed defect',
    '7.0': 'Rreversable defect'
})


In [4]:
# define a column, 'heart_disease', based on the 'diagnosis' column and drop 'diagnosis'
hd_data["heart_disease"] = hd_data["diagnosis"].apply(
    lambda x: "undiagnosed" if (x == 0) else "diagnosed")
hd_data = hd_data.drop(columns=["diagnosis"])

hd_data.head()

age     sex        chest_pain  resting_blood_pressure  cholesterol  \
0  63.0    male    Typical angina                   145.0        233.0   
1  67.0    male      Asymptomatic                   160.0        286.0   
2  67.0    male      Asymptomatic                   120.0        229.0   
3  37.0    male  Non-anginal pain                   130.0        250.0   
4  41.0  female   Atypical angina                   130.0        204.0   

   fasting_blood_sugar  resting_electrocardiographic_results  max_heart_rate  \
0                 True                                   2.0           150.0   
1                False                                   2.0           108.0   
2                False                                   2.0           129.0   
3                False                                   0.0           187.0   
4                False                                   2.0           172.0   

  exercise_induced_angina  st_depression_exercise     slope_st  major_vessels  \
0                      No                     2.3  Downsloping            0.0   
1                     Yes                     1.5         Flat            3.0   
2                     Yes                     2.6         Flat            2.0   
3                      No                     3.5  Downsloping            0.0   
4                      No                     1.4    Upsloping            0.0   

                 thal heart_disease  
0        Fixed defect   undiagnosed  
1              Normal     diagnosed  
2  Rreversable defect     diagnosed  
3              Normal   undiagnosed  
4              Normal   undiagnosed

Now we have a 14-column dataset that contains six numerical variables and eight categorical variables. 

| # | Variable | Description | Value | Possible Correlation |
|---|----------|-------------|-------|----------------------|
| 1 | age | Individual's age | numerical | Yes |
| 2 | sex | Individual's sex | male or female | Yes |
| 3 | chest_pain | Chest pain type | Typical angina, Atypical angina, Non-anginal pain, or Asymptomatic | Yes |
| 4 | resting_blood_pressure | Resting blood pressure (in mm Hg on admission to the hospital) | numerical | worth checking again  |
| 5 | cholesterol | Serum cholestoral in mg/dL | numerical | need re-analysis  |
| 6 | fasting_blood_sugar | Fasting blood sugar > 120 mg/dL | True or False | No |
| 7 | resting_electrocardiographic_results | Resting electrocardiographic results | <ul><li>0: normal</li><li>1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)</li><li>2: showing probable or definite left ventricular hypertrophy by Estes' criteria</li></ul> | No |
| 8 | max_heart_rate | Maximum heart rate achieved | numerical | Yes |
| 9 | exercise_induced_angina | Exercise induced angina | Yes or No | Yes |
| 10 | st_depression_exercise | ST depression induced by exercise relative to rest | numerical | Yes |
| 11 | slope_st | The slope of the peak exercise ST segment | Upsloping, Flat, or Downsloping | Yes (but too technical) |
| 12 | major_vessels | Number of major vessels (0-3) colored by flourosopy | numerical | Yes |
| 13 | thal | Thalassemia | Normal, Fixed defect, or Reversable defect | Yes |
| 14 | heart_disease | Diagnosis of presence of heart disease | diagnosed or undiagnosed | NA |

### Scaling and Centering

We will show some statistics for all six numeric variables in the original data.

In [5]:
age_stat = hd_data.agg({'age': ['mean', 'median', 'min', 'max']}).round(decimals=1)
rbp_stat = hd_data.agg({'resting_blood_pressure': ['mean', 'median', 'min', 'max']}).round(decimals=2)
chol_stat = hd_data.agg({'cholesterol': ['mean', 'median', 'min', 'max']}).round(decimals=2)
max_hr_stat = hd_data.agg({'max_heart_rate': ['mean', 'median', 'min', 'max']}).round(decimals=2)
st_depression_exercise = hd_data.agg({'st_depression_exercise': ['mean', 'median', 'min', 'max']}).round(decimals=2)
major_vessels = hd_data.agg({'major_vessels': ['mean', 'median', 'min', 'max']}).round(decimals=2)

num_stat = pd.concat([age_stat, rbp_stat, chol_stat, max_hr_stat, st_depression_exercise, major_vessels], axis=1)
num_stat.transpose()

mean  median    min    max
age                      54.50    56.0   29.0   77.0
resting_blood_pressure  131.69   130.0   94.0  200.0
cholesterol             247.35   243.0  126.0  564.0
max_heart_rate          149.60   153.0   71.0  202.0
st_depression_exercise    1.06     0.8    0.0    6.2
major_vessels             0.68     0.0    0.0    3.0

We can see the variation of scales in the above variables, so we will scale these values before using them for our analysis with the K nearest neighbours algorithms.

In [6]:
hd_preprocessor = make_column_transformer(
    (StandardScaler(), [
        "age",
        "resting_blood_pressure",
        "cholesterol",
        "max_heart_rate",
        "st_depression_exercise",
        "major_vessels"
    ]),
    ("passthrough", [
        "sex",
        "chest_pain",
        "fasting_blood_sugar",
        "resting_electrocardiographic_results",
        "exercise_induced_angina",
        "slope_st", 
        "thal",
        "heart_disease"
    ]),
)

hd_scaled = pd.DataFrame(
    hd_preprocessor.fit_transform(hd_data),
    columns=[
        "age",
        "resting_blood_pressure",
        "cholesterol",
        "max_heart_rate",
        "st_depression_exercise",
        "major_vessels",
        "sex",
        "chest_pain",
        "fasting_blood_sugar",
        "resting_electrocardiographic_results",
        "exercise_induced_angina",
        "slope_st", 
        "thal",
        "heart_disease"
    ],
)
hd_scaled.head()

age resting_blood_pressure cholesterol max_heart_rate  \
0  0.936181                0.75038   -0.276443       0.017494   
1  1.378929               1.596266    0.744555      -1.816334   
2  1.378929              -0.659431     -0.3535       -0.89942   
3  -1.94168              -0.095506    0.051047        1.63301   
4 -1.498933              -0.095506   -0.835103       0.978071   

  st_depression_exercise major_vessels     sex        chest_pain  \
0               1.068965     -0.721976    male    Typical angina   
1               0.381773      2.478425    male      Asymptomatic   
2               1.326662      1.411625    male      Asymptomatic   
3               2.099753     -0.721976    male  Non-anginal pain   
4               0.295874     -0.721976  female   Atypical angina   

  fasting_blood_sugar resting_electrocardiographic_results  \
0                True                                  2.0   
1               False                                  2.0   
2               False                                  2.0   
3               False                                  0.0   
4               False                                  2.0   

  exercise_induced_angina     slope_st                thal heart_disease  
0                      No  Downsloping        Fixed defect   undiagnosed  
1                     Yes         Flat              Normal     diagnosed  
2                     Yes         Flat  Rreversable defect     diagnosed  
3                      No  Downsloping              Normal   undiagnosed  
4                      No    Upsloping              Normal   undiagnosed

### Training Data and Test Data

We will obtain a training data and a test data from the `hd_scaled` data and *use only the training data* from now on.

In [7]:
from sklearn.model_selection import train_test_split

hd_train, hd_test = train_test_split(hd_scaled, test_size=0.25, random_state=123)